### PR TITLE
add ci job to pre-build osx and Linux multi-arch

### LINKED ISSUE #3: Pre-builds for more platforms /  architectures
**Is your feature request related to a problem? Please describe.**
Looks like currently there are pre-builds happening for linux via: 
https://github.com/matrix-org/matrix-rust-sdk-bindings/blob/main/crates/matrix-sdk-crypto-nodejs/release/Dockerfile.linux#L34

But for others it utilizes a post install hook:
https://github.com/matrix-org/matrix-rust-sdk-bindings/blob/main/crates/matrix-sdk-crypto-nodejs/check-exists.js#L7

Which then means you need to have rust, cargo and numerous other dependencies installed.

**Describe the solution you'd like**

Looks like the current CI builds for Mac and linux.  I'd propose utilizing the CI instead to do the build of the artifacts. 

The hard part is need to then release to npm from the CI.  Or..  upload as artifacts and make the post install fetch from those build artifacts if not present.

Honestly not sure on the best solution.

**Describe alternatives you've considered**

**Additional context**

In my particular case.  The check-exists fails regardless because we utilize yarn workspaces.  Something about having yarn workspace plugin configured causes the `yarn build:release` from inside the check-exists.js to fail with an error like:

```
# This file contains the result of Yarn building a package (@turt2live/matrix-sdk-crypto-nodejs@npm:0.1.0-beta.10)
# Script name: postinstall

[31m[1mInternal Error[22m[39m: @turt2live/matrix-sdk-crypto-nodejs@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
    at tF.getCandidates (/Users/aaron/go/src/github.com/RocketChat/Rocket.Chat/.yarn/releases/yarn-3.2.0.cjs:437:4480)
    at wd.getCandidates (/Users/aaron/go/src/github.com/RocketChat/Rocket.Chat/.yarn/releases/yarn-3.2.0.cjs:395:1281)
    at /Users/aaron/go/src/github.com/RocketChat/Rocket.Chat/.yarn/releases/yarn-3.2.0.cjs:441:7765
    at Pg (/Users/aaron/go/src/github.com/RocketChat/Rocket.Chat/.yarn/releases/yarn-3.2.0.cjs:394:11098)
    at le (/Users/aaron/go/src/github.com/RocketChat/Rocket.Chat/.yarn/releases/yarn-3.2.0.cjs:441:7745)
```

you can reproduce by cloning: https://github.com/RocketChat/Rocket.Chat on a Mac and doing `yarn add matrix-appservice-bridge` or directly this package `yarn add @turt2live/matrix-sdk-crypto-nodejs`.

I believe its related to yarn workspaces as if I copy the .yarn folder and .yarnrc.yml to a new folder and try the same error happens.

Current work around is go into node_modules/@turt2live/matrix-sdk-crypto-nodejs and run `npm run build:debug` or `npm run build:release`

### PR DESCRIPTION
**Pull request by geekgonecrazy**
Closes #3 

I have a failing job included for musl - needed to compile on alpine. 

I'm going to keep battling a bit.. might be a lost cause.

But this will compile both x64 and arm64 for osx as well as i686 and x64 for linux and given credentials can ship to npm registry automatically.

Example of the CI running: https://github.com/geekgonecrazy/matrix-rust-sdk-bindings/actions/runs/2222780302

### CODE REVIEW

File: .github/workflows/build-npm-and-release.yml [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+          ls -l lib
+ 
+          if git log -1 --pretty=%B | grep "^[0-9]\+\.[0-9]\+\.[0-9]\+$";
+          then
+            echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
+            npm publish --access public
+          elif git log -1 --pretty=%B | grep "^[0-9]\+\.[0-9]\+\.[0-9]\+";
+          then
+            echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
+            npm publish --tag next --access public
```
Comment (geekgonecrazy) [inline .github/workflows/build-npm-and-release.yml]: Pulled this bit from the napi example project. So probably wouldn't need this bit

File: .github/workflows/build-npm-and-release.yml [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+          docker tag ghcr.io/napi-rs/napi-rs/nodejs-rust:lts-alpine builder
+
+          docker run --rm -v $(pwd):/build -w /build builder yarn --ignore-scripts 
+          docker run --rm -v $(pwd):/build -w /build builder yarn build:linux:x64:musl
+      - name: Upload artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: index.linux-x64-musl.node
+          path: crates/matrix-sdk-crypto-nodejs/lib/index.linux-x64-musl.node
+          if-no-files-found: error
```
Comment (geekgonecrazy) [inline .github/workflows/build-npm-and-release.yml]: i've left this job so that publish doesn't depend on it.  This is the one building on alpine.

Seems to be a pretty consistent error:

```
      Run Build Command(s):/usr/bin/make -f Makefile cmTC_d9c2d/fast && /usr/bin/make  -f CMakeFiles/cmTC_d9c2d.dir/build.make CMakeFiles/cmTC_d9c2d.dir/build
      make[1]: Entering directory '/build/target/x86_64-unknown-linux-musl/release/build/olm-sys-b3fdfb386297bf25/out/build/CMakeFiles/CMakeTmp'
      Building CXX object CMakeFiles/cmTC_d9c2d.dir/testCXXCompiler.cxx.o
      /usr/bin/clang++   -ffunction-sections -fdata-sections -fPIC --target=x86_64-unknown-linux-musl  -MD -MT CMakeFiles/cmTC_d9c2d.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_d9c2d.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_d9c2d.dir/testCXXCompiler.cxx.o -c /build/target/x86_64-unknown-linux-musl/release/build/olm-sys-b3fdfb386297bf25/out/build/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
      Linking CXX executable cmTC_d9c2d
      /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_d9c2d.dir/link.txt --verbose=1
      /usr/bin/clang++  -ffunction-sections -fdata-sections -fPIC --target=x86_64-unknown-linux-musl  CMakeFiles/cmTC_d9c2d.dir/testCXXCompiler.cxx.o -o cmTC_d9c2d 
      /usr/bin/ld: cannot find crtbeginS.o: No such file or directory
      /usr/bin/ld: cannot find -lgcc: No such file or directory
      clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
      make[1]: *** [CMakeFiles/cmTC_d9c2d.dir/build.make:100: cmTC_d9c2d] Error 1
      make[1]: Leaving directory '/build/target/x86_64-unknown-linux-musl/release/build/olm-sys-b3fdfb386297bf25/out/build/CMakeFiles/CMakeTmp'
      make: *** [Makefile:127: cmTC_d9c2d/fast] Error 2
```

Comment (geekgonecrazy) [reply .github/workflows/build-npm-and-release.yml]: Disregard.. persistence pays off

**Comment by geekgonecrazy** [comment]
https://github.com/geekgonecrazy/matrix-rust-sdk-bindings/actions/runs/2229035529 alright shouldn't need to mess with it any more.  Here is an example of it finishing to publish.  

We temporarily created a fork copy of the npm packages which can be found in the PR that referred above.

**Review by turt2live — COMMENTED** [review]
Overall this is really great to see. It was always the intention to add more prebuilds, though doing so completely with CI is a bit interesting. 

This would be our first project to publish directly from GHA, so there's extra concern which goes into this. The original intent of the build pipeline we've made is to allow a developer on a release machine run `yarn publish` and it'll take care of the pre-builds and such, so environmental dependencies are entirely valid. 

It's true though that without prebuilds we end up in a case where downstream developers need the same pipeline, which isn't great. The plan was to provide prebuilds for Mac, Windows, and Linux before pushing this library out widely :) 

I'll have our team look into safely using the npm token, but in the meantime if you have resources for how people manage npm releases managed by a CI pipeline it'd be appreciated. Of concern would be the awkward dance for git tagging: `yarn publish` currently does a lot of the heavy lifting for us, including picking a new version number, but we'd lose that automation as part of requiring a CI pipeline.

**Comment by geekgonecrazy** [comment]
Sorry for the delayed response here.

Your approach sounds exactly like what my initial thought was.  Having prebuilts available even remotely that is fetched during packaging would to me work just fine.  I just went the route of trying to automate end to end because seemed the quickest for me.  Little did I know the trouble I would run into just getting some of the builds to even be reproducible consistently in a CI 🙈

--- 

As to the using the CI.  To me I like this approach because makes the builds repeatable and no one has to worry about getting things just right on a bunch of machines and then combining them manually together for a release.

We do utilize github CI a lot in Rocket.Chat and have it publish to docker, npm, snap store etc.  So we've put a good bit of trust in it.  But we feel it's important to be able to automate these sort of things.

---

Some notes about this PR.  Since this PR.. The alpine build does not work.  While it compiles and everything seems great.. when trying to load on alpine it depends on a memcpy that is not in the musl glibc compatibilty package.  So it might should be excluded from this build.

I've spent many hours banging my head against the wall for it.  Tried building on alpine, introducing newer versions of glibc compat compiled from source for alpine...   Also tried different compile time options to try and make use musl-cc instead of gcc or musl-gcc (apparently now days its just musl-gcc). 

For ubuntu / centos etc.. it seems that need to target an older version of glibc... as does work on newest ubuntu it does not work on 16.04 and centos at all.  

It's possible that targeting an older version of glibc would use an instruction set that is in the musl compatibilty layer.

The problem with that so far is that in trying to automate that I tried on Ubuntu 16.04 but it then needs a newer version of CMake for rust to be able to compile in parallel.  So I kind of gave up for now on that.

---
So in summary its' definitely missing the bow tied on top. Feel free to take what you will here.  The mac builds seem to work great so far so at very least that seems to be a good addition.

**Comment by gnunicorn** [comment]
Our action plan is release a complete rewrite of the nodejs-bindings based on the latest crypto-ffi crates, using vodozemac. So in that case we wouldn't need CMAKE and all that stuff anymore but it would only built using rust. Would that alleviate your build pains, @geekgonecrazy ?

ref https://github.com/matrix-org/matrix-rust-sdk/issues/699

**Comment by geekgonecrazy** [comment]
I haven't fully understood what all of the plan means yet.  Based on what I heard in the matrix live podcast.  It sounds like the plan is to rely more heavily on it.  And wasm is at play?  I haven't really ever used wasm with node.  Not sure if arch specific builds are needed?  I kind of assume so.

**If** its still loading native code still it will probably still going to need to have prebuilt for each platform and architecture right?  So I don't think it would actually fix the main problem. 

The main problem is when the package is installed that computer has to have the whole rust toolchain just to be able to install the module.  Which for normal devs and server admins they don't have this preinstalled.  

The normal flow for these devs and end users is:
1. Download the release bundle
2. npm install
3. run

If we don't ship pre-built then the flow becomes
1. Download release bundle
2. npm install
3. it fails..
4. go find an issue or communication somewhere explaining what I have to do
5. have difficulties installing for chosen platform or distro (because lets face it this is just what happens)
6. Hopefully then run it and it work

**Comment by geekgonecrazy** [comment]
For alpine specifically and older versions of ubuntu... this might work.  As the module might not depend so heavily on glibc? I don't know anything about the crypto-ffi crates so can't really be sure

**Comment by poljar** [comment]
So the thing is that we can't in good conscience let CI take over the release process of such a security critical component.

The plan is to eventually provide pre-built binaries for a bunch of bindings for the `matrix-sdk-crypto` crate, but we don't have yet the infrastructure set up.

In the meantime, the build has been simplified since it includes only Rust code nowadays. `cargo build` should just work for compilation as well as for cross-compilation.

> I do see wasm mentioned there. I'm not sure if that means that the need for platform specific builds goes away?

It does if you're in browser land, the storage implementation for WASM uses IndexedDB, Node bindings are likely still wanted, to get the performance gains of a native executable, and to avoid having to mess around with IndexedDB shims for node.

**Comment by geekgonecrazy** [comment]
> So the thing is that we can't in good conscience let CI take over the release process of such a security critical component.

I mean end of the day this is y'all's choice as the maintainers. If it makes sense for y'all to maintain infrastructure and manually build things instead of letting a CI build.  That's definitely your choice.  

As long as the releases end up being made with the prebuilts included for all the other platforms.  Our project and end users will be happy.

Either way Looking forward to when the builds are pure rust and can successfully crosscompile.  Keeping an eye on that development.  

I'll go ahead and close this PR out
